### PR TITLE
GLUE tasks for bert

### PR DESCRIPTION
**Pull request by ekmb**
Signed-off-by: ekmb <e.bakhturina@gmail.com>

### CODE REVIEW

File: collections/nemo_nlp/nemo_nlp/data/datasets/__init__.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
                                 BertJointIntentSlotInferDataset)
 from .language_modeling import LanguageModelingDataset
 from .ner import BertNERDataset
+from .glue import GLUEDataset
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/__init__.py]: arrange import alphabetically

File: collections/nemo_nlp/nemo_nlp/data/datasets/glue.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+import numpy as np
+from torch.utils.data import Dataset
+
+from nemo.utils.exp_logging import get_logger
+
+logger = get_logger('')
+
+class GLUEDataset(Dataset):
+    def __init__(self, tokenizer, data_dir, max_seq_length, processor,
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/glue.py]: each argument should be in its own line

File: collections/nemo_nlp/nemo_nlp/data/datasets/glue.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            special_tokens_count += 2 if bos_token else 0
+            special_tokens_count += 1 if cls_token else 0
+            _truncate_seq_pair(tokens_a, tokens_b, max_seq_length - \
+                               special_tokens_count)
+        else:
+            special_tokens_count = 1 if eos_token else 0
+            special_tokens_count += 1 if sep_token_extra else 0
+            special_tokens_count += 1 if bos_token else 0
+            if len(tokens_a) > max_seq_length - special_tokens_count:
+                tokens_a = tokens_a[:(max_seq_length - special_tokens_count)]
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/glue.py]: this () is unnecessary

**Comment by okuchaiev** [comment]
a lot of PEP8 errors (which is why CI failed). Please fix those before the merge. You can check for PEP8 locally like this:
``
pycodestyle . --exclude=./scripts/get_librispeech_data.py,./scripts/process_beam_dump.py,./tests/other/jasper.py,./tests/other/jasper_zero_dl.py,./collections/nemo_nlp/nemo_nlp/utils/metrics/sacrebleu.py,./collections/nemo_nlp/nemo_nlp/utils/metrics/fairseq_tokenizer.py,./nemo/setup.py,./docs/sources/source/conf.py,./docs/sources/source/tutorials/infer.py,./docs/sources/source/tutorials/test.py,./collections/nemo_nlp/build
``

File: collections/nemo_nlp/nemo_nlp/data/datasets/utils.py [hunk 4]
```diff
@@ -1,11 +1,16 @@
+from __future__ import absolute_import, division, print_function
```
Comment (blisc) [inline collections/nemo_nlp/nemo_nlp/data/datasets/utils.py]: Let's remove this import

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/glue.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    if output_mode == 'classification':
+        preds = list(np.argmax(np.asarray(logits_lists), 1))
+    elif output_mode == 'regression':
+        preds = list(np.squeeze(np.asarray(logits_lists)))
+    global_vars["all_preds"].extend(preds)
+    global_vars["all_labels"].extend(labels_lists)
+
+
+def list2str(l):
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/utils/callbacks/glue.py]: this is a function in data/datasets/utils that you can call

File: examples/nlp/run_glue_with_BERT.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+    loss = glue_loss(logits=logits, labels=labels)
+
+    steps_per_epoch = len(data_layer) // (batch_size * num_gpus)
+    return loss, steps_per_epoch, data_layer, [logits, labels]
+
+
+train_loss, steps_per_epoch, _, _ = create_pipeline(train_dataset)
+_, _, eval_data_layer, eval_tensors = \
+                               create_pipeline(eval_datasets[0])
```
Comment (chiphuyen) [inline examples/nlp/run_glue_with_BERT.py]: for tasks with more than one eval datasets, do we only evaluate the first dataset?

File: examples/nlp/run_glue_with_BERT.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        x, y, eval_data_layer, output_mode),
+    user_epochs_done_callback=lambda x: eval_epochs_done_callback(
+        x, args.work_dir, eval_task_names[0]),
+    tb_writer=nf.tb_writer,
+    eval_step=steps_per_epoch)]
+
+# create additional callback and data layer for MNLI mismatched dev set
+if args.task_name == 'mnli':
+    _, _, eval_data_layer_mm, eval_tensors_mm = \
+                               create_pipeline(eval_datasets[1])
```
Comment (chiphuyen) [inline examples/nlp/run_glue_with_BERT.py]: we should find a way to be able to evaluate all eval datasets, not just hardcode 0 and 1

File: nemo/nemo/backends/pytorch/common/losses.py [hunk 8]
```diff
@@ -137,3 +137,30 @@ def _loss_function(self,
...
+            "loss": NeuralType(None)
+        }
+        return input_ports, output_ports
+
+    def __init__(self, **kwargs):
+        LossNM.__init__(self, **kwargs)
+        self._criterion = nn.MSELoss()
+
+    def _loss_function(self, logits, labels):
+        loss = self._criterion(logits.view(-1), labels)
```
Comment (chiphuyen) [inline nemo/nemo/backends/pytorch/common/losses.py]: MSELoss takes in predictions and labels, not logits and labels

File: examples/nlp/run_glue_with_BERT.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+                                   add_time_to_log_dir=True)
+
+if args.bert_checkpoint is None:
+    """ Use this if you're using a standard BERT model.
+    To see the list of pretrained models, call:
+    nemo_nlp.huggingface.BERT.list_pretrained_models()
+    """
+    tokenizer = NemoBertTokenizer(args.pretrained_bert_model)
+    model = nemo_nlp.huggingface.BERT(
+        pretrained_model_name=args.pretrained_bert_model, factory=nf)
```
Comment (chiphuyen) [inline examples/nlp/run_glue_with_BERT.py]: we no longer need factory=nf for any NeMo modules

File: examples/nlp/run_glue_with_BERT.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+from nemo_nlp.data.datasets.utils import processors, output_modes
+
+parser = argparse.ArgumentParser(description="GLUE_with_pretrained_BERT")
+
+# Parsing arguments
+parser.add_argument("--data_dir", default='COLA', type=str, required=True,
+                    help="The input data dir. Should contain the .tsv    \
+                    files (or other data files) for the task.")
+parser.add_argument("--task_name", default="CoLA", type=str, required=True,
```
Comment (chiphuyen) [inline examples/nlp/run_glue_with_BERT.py]: error when user enters an invalid task name

File: collections/nemo_nlp/nemo_nlp/data/datasets/glue.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+        else:
+            input_ids = input_ids + ([pad_token_id] * padding_length)
+            input_mask = input_mask + \
+                ([0 if mask_padding_with_zero else 1] * padding_length)
+            segment_ids = segment_ids + \
+                ([pad_token_segment_id] * padding_length)
+        if len(input_ids) != max_seq_length:
+            raise ValueError("inpud_ids must be of length max_seq_length")
+        if len(input_mask) != max_seq_length:
+            raise ValueError("inpud_mask must be of length max_seq_length")
```
Comment (redoctopus) [inline collections/nemo_nlp/nemo_nlp/data/datasets/glue.py]: Minor error message typo ("inpud" vs "input")

File: collections/nemo_nlp/nemo_nlp/data/data_layers.py [hunk 12]
```diff
@@ -414,3 +416,119 @@ def dataset(self):
...
+                          'processor': processor,
+                          'evaluate': evaluate,
+                          'token_params': token_params,
+                          'tokenizer': tokenizer,
+                          'max_seq_length': max_seq_length}
+
+        super().__init__(dataset_type, dataset_params, **kwargs)
+
+
+class GlueDataLayerRegression(TextDataLayer):
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/data_layers.py]: we can leave it as it is for now but i think separating GlueDataLayerClassification and GlueDataLayerRegression only makes me more confusing to users.

File: collections/nemo_nlp/nemo_nlp/data/datasets/glue.py [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+
+    def __len__(self):
+        return len(self.features)
+
+    def __getitem__(self, idx):
+        feature = self.features[idx]
+        return np.array(feature.input_ids), \
+            np.array(feature.segment_ids),  \
+            np.array(feature.input_mask, dtype=np.float32), \
+            np.array(feature.label_id)
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/glue.py]: you can use () for alignment
return (item1,
           item2)

File: collections/nemo_nlp/nemo_nlp/data/datasets/glue.py [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+            logger.info("*** Example ***")
+            logger.info("guid: %s" % (example.guid))
+            logger.info(
+                "tokens: %s" % " ".join([str(x) for x in tokens]))
+            logger.info(
+                "input_ids: %s" % " ".join([str(x) for x in input_ids]))
+            logger.info(
+                "input_mask: %s" % " ".join([str(x) for x in input_mask]))
+            logger.info(
+                "segment_ids: %s" % " ".join([str(x) for x in segment_ids]))
```
Comment (chiphuyen) [inline collections/nemo_nlp/nemo_nlp/data/datasets/glue.py]: use map for this. e.g.:
"segment_ids: %s" % " ".join(list(map(str, segment_ids)))

File: collections/nemo_nlp/nemo_nlp/utils/callbacks/glue.py (hunk 5 above)
Comment (chiphuyen) [reply collections/nemo_nlp/nemo_nlp/utils/callbacks/glue.py]: @ekmb

File: examples/nlp/run_glue_with_BERT.py (hunk 6 above)
Comment (ekmb) [reply examples/nlp/run_glue_with_BERT.py]: because MNLI has 2 dev sets: matched and mismatched
